Strassen

amkatrutsa opened the pull request:
Add Strassen algorithm for matmul

ovgeorge committed:
Strassen: added tile-based implementation

Divide tensors into quaters by tile and conduct opertations between tiles using already implemented methods.
Does not work with leftover tile (i.e. if tiles of the tensor differ in their sizes).

ovgeorge committed:
Strassen: python wrappers

ovgeorge committed:
Strassen: tests for different configuration of input parameters

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+        if (
+            np.linalg.norm(dst_C[:, :, i] - src_C[:, :, i])
+            / np.linalg.norm(src_C[:, :, i])
+            > 1e-4
```
It's better to use `assert_array_equal` or `assert_allclose` from `numpy.testing` (see [docs][1] for details).

[1]: https://numpy.org/doc/stable/reference/routines.testing.html

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+
+    C.to_array(dst_C)
+    strassen[dtype](alpha, tA, A, tB, B, beta, C, 1, 1, 0)
+    C.to_array(dst_C)
```
@Muxas We need to rework at this point. It's common point to use Buffer Protocol for interop.

```python
A = nntile_tensor_create_routine(src_A)  # From NumPy array.
dst_C = np.asarray(C)  # To NumPy array.
```

daskol replied:
#28 seems to be wrong issue.

Muxas replied:
Indeed, correct one is #39

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+    strassen[dtype](alpha, tA, A, tB, B, beta, C, 1, 1, 0)
+    C.to_array(dst_C)
+
+    nntile.starpu.wait_for_all()
```
@Muxas FYI People called `.block_until_ready()` a routine of such kind.

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+# Define list of tested types
+dtypes = [np.float32, np.float64]
+# Define mapping between numpy and nntile types
+Tensor = {np.float32: nntile.tensor.Tensor_fp32, np.float64: nntile.tensor.Tensor_fp64}
```
@Muxas We need high-level array creation routine to avoid boilerplate everywhere.

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+# @author Aleksandr Mikhalev
+# @date 2023-12-18
+
+# All necesary imports
```
Remove excessive comment please.

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+
+
+# Helper function returns bool value true if test passes
+def helper(dtype, tA, tB, matrix_shape, shared_size, tile_size, alpha, beta, batch=3):
```
Instead of describing function with a comment, write a oneliner in a docstring as follows.

def helper(dtype, tA, tB, matrix_shape, shared_size, tile_size, alpha, beta, batch=3):
    ""Helper function returns bool value true if test passes."""

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+
+
+# Repeat tests for different configurations
+def tests():
```
@Muxas What are our plans for testing? We should avoid this boilerplate with `pytest` and parametrized test as follows.

```python
@pytest.mark.parametrized('dtype', [np.float32, np.float64])
@pytest.mark.parametrized('trans', [False, True])
def test_strassen(dtype, trans, ...):
    ...
```

But I'm not sure that this will be work properly since multiple test could be run and multiple runtimes can be created. It might break something. Thus we at least should add `pytest-timeout` to fail faster.

daskol commented on wrappers/python/nntile/nntile_core.cc:
```diff
@@ -739,6 +739,10 @@ void def_mod_tensor(py::module_ &m)
+
+    m.def("strassen_fp64", &strassen<fp64_t, fp64_t>);
+    m.def("strassen_fp32", &strassen<fp32_t, fp32_t>);
+    // m.def("strassen_fp16", &strassen<fp16_t, fp32_t>);
```
Do not leave dead code in comment please.

daskol commented on wrappers/python/tests/nntile_core/test_tensor_strassen.py:
```diff
@@ -0,0 +1,113 @@
+# Test for tensor::strassen<T> Python wrapper
+#
+# @version 1.0.0
+# @author Aleksandr Mikhalev
```
Who is the author?

@Muxas May be we fill author as **NNTile Team** from now on?

daskol replied:
How will maintain these filed and watch for their consistency? Should one add oneself to authors during project-wide refactoring?

Muxas replied:
Example of filling in authors and organisations was taken from reference BLAS source code. @daskol please provide examples from other well known open source project to start a discuss.

daskol commented on include/nntile/tensor/strassen.hh:
```diff
@@ -0,0 +1,41 @@
+namespace nntile
+{
+namespace tensor
+{
```
It is better to use shorter syntax.

```cpp
namespace nntile::tensor {
    ...
} // nntile::tensor
```

daskol replied:
It seems that CUDA supports C++20 since 12.4.

ovgeorge committed:
Strassen: added perf script

ovgeorge committed:
Use strassen in linear layer

By default strassen will be used.
If exception is thrown by strassen_check, GEMM is used instead.

ovgeorge committed:
Apply black to linear.py

ovgeorge committed:
Use strassen in linear layer

By default strassen will be used.
If exception would be thrown by strassen_check due to leftover tile, GEMM is used instead.
